### ISSUE TITLE
[PAID] [$250] [Workspace Feeds] App crashes after creating new WS and selecting start over

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#49447]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** v9.0.38-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Unable to validate in Prod because of Bank Setup
**Issue reported by:** Applause Internal Team

## Action Performed:
1. Open the app and log in
2. Create a workspace
3. Navigate to More features and enable Expensify card
4. Tap Issue new card > select Connect online with Plaid and add VBA
5. Create a new workspace and enable Expensify card in More features
6. Tap Issue new card > + Add a new bank account and select Start over

## Expected Result:
The options to add a bank account using Plaid or entering manually are displayed

## Actual Result:
The app crashes

## Workaround:
Unknown

## Platforms:

- [x] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

[1909_1.txt](https://github.com/user-attachments/files/17056552/1909_1.txt)

https://github.com/user-attachments/assets/3f86ea4b-522f-4f8e-bbe8-f0d74066cb52

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021836823617385889668</li>
        <li>Upwork Job ID: 1836823617385889668</li>
        <li>Last Price Increase: 2024-10-24</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @CortneyOfstad</details>

**Comment by IuliiaHerets** [comment]
We think that this bug might be related to #wave-collect - Release 2

**Comment by Julesssss** [comment]
I'm just trying to figure out why this isn't testable in prod. 

> Unable to validate in Prod because of Bank Setup

@IuliiaHerets is it possible to test in prod with a different account? Or is this due to product functionality not being available?

**Comment by trjExpensify** [comment]
Probably because the Plaid sandbox that loads for the test VBBA credentials only works on staging or something.

**Comment by trjExpensify** [comment]
>I can investigate this issue

Please do though, @koko57!

**Comment by koko57** [comment]
waiting for my android builds... On the web I don't see the problem after clicking "Start over". But I think we shouldn't show the "Start over" option as we already added an account, especially for a new workspace, am I right?

**Comment by mountiny** [comment]
I can sheppard this one, its definitely weird. in web there would probably be some console error.

**Comment by m-natarajan** [comment]
Able to reproduce this issue
email used: applausetester+091924pm1@applause.expensifail.com (Precondition: New account)
Device details:  Pixel 6 / Android 14

https://github.com/user-attachments/assets/d49e3634-781d-49cf-a93f-aa709d7ecf2d

**Comment by mountiny** [comment]
I think it will be this crashlog

```
 Fatal Exception: com.facebook.react.common.JavascriptException: Error: Missing bankAccountID when attempting to reset free plan bank account, js engine: hermes, stack:
resetFreePlanBankAccount@1:5269051
t5@1:8671650
t16@1:4785522
onPress@1:4245614
t23@1:4218512
anonymous@1:4219664
_performTransitionSideEffects@1:791974
_receiveSignal@1:791562
onResponderRelease@1:790391
executeDispatch@1:1250761
executeDispatchesAndReleaseTopLevel@1:1256047
forEachAccumulated@1:1252277
anonymous@1:1256485
batchedUpdatesImpl@1:1317425
batchedUpdates$1@1:1255971
dispatchEvent@1:1256250

       at com.facebook.react.modules.core.ExceptionsManagerModule.reportException(ExceptionsManagerModule.java:65)
       at com.facebook.jni.NativeRunnable.run(NativeRunnable.java)
       at android.os.Handler.handleCallback(Handler.java:959)
       at android.os.Handler.dispatchMessage(Handler.java:100)
       at com.facebook.react.bridge.queue.MessageQueueThreadHandler.dispatchMessage(MessageQueueThreadHandler.java:27)
       at android.os.Looper.loopOnce(Looper.java:232)
       at android.os.Looper.loop(Looper.java:317)
       at com.facebook.react.bridge.queue.MessageQueueThreadImpl$4.run(MessageQueueThreadImpl.java:235)
       at java.lang.Thread.run(Thread.java:1012)
```

**Comment by grgia** [comment]
To me this seems like something that exists on prod and the root cause is likely that we’re just passing ONYXKEYS.REIMBURSEMENT_ACCOUNT here even though it’s null or outdated

**Comment by grgia** [comment]
Confirmed that's the root cause by testing starting one without plaid flow 
<img width="518" alt="Screenshot 2024-09-19 at 6 41 06 PM" src="https://github.com/user-attachments/assets/345ed0da-af86-47e0-921c-01888bfed783">

**Comment by grgia** [comment]
# Overview of Solution:
We need to store some identifier key for the bank account as well (policy#, accountID, etc) so we can ensure that we're not using stale bank account data from a different flow

cc @BrtqKr as you implemented the original key

**Comment by grgia** [comment]
Demoting, because the above test was done on prod :)

**Comment by trjExpensify** [comment]
This still `External`?

**Comment by koko57** [comment]
@grgia @mountiny but anyway, shouldn't we show a new VBBA page instead of Continue modal for a new workspace?
<img width="374" alt="Screenshot 2024-09-20 at 18 07 34" src="https://github.com/user-attachments/assets/37e39786-4d39-48e3-a779-73cd0d970a43">

If we had a bankAccountID and we would choose "Start over" will it conflict with adding the account for the previous workspace?

**Comment by trjExpensify** [comment]
@shawnborton I could have sworn there was a version of this that has a section at the bottom that says "Use existing account" that lists any VBBA the admin has added to their account can be selected to use. Am I making that up? 🤔

**Comment by koko57** [comment]
I tried to reproduce it once again and I was able to reproduce it on web. 
<img width="1419" alt="Screenshot 2024-09-20 at 18 36 05" src="https://github.com/user-attachments/assets/18d825b2-ed11-40a5-8c59-8e548914cb95">


I noticed that first we set `shouldShowContinueSetupButton` to true here https://github.com/Expensify/App/blob/d009d8ed4830c7d0562c524c991b8a688e2b4869/src/pages/ReimbursementAccount/ReimbursementAccountPage.tsx#L208
because `getShouldShowContinueSetupButtonInitialValue` returns true, but in the meantime we send `OpenReimbursementAccountPage` request and we end up with this data 
<img width="394" alt="Screenshot 2024-09-20 at 18 47 32" src="https://github.com/user-attachments/assets/48d937a1-189f-401f-bb8a-d5708e3bb838">, but the page doesn't rerender to show the new VBBA flow page but the Continue modal is still there.

So as we don't have bankAccountID there we end up throwing this error https://github.com/Expensify/App/blob/d009d8ed4830c7d0562c524c991b8a688e2b4869/src/libs/actions/ReimbursementAccount/resetFreePlanBankAccount.ts#L22

Maybe we should handle this error differently though?

**Comment by shawnborton** [comment]
> Am I making that up? 🤔

That sounds really familiar to me too, I could have sworn that was the case as well.

**Comment by mountiny** [comment]
> I could have sworn there was a version of this that has a section at the bottom that says "Use existing account" that lists any VBBA the admin has added to their account can be selected to use. Am I making that up? 🤔

Yeah same there should be a list of eligible bank accounts to pick from. @koko57 can you please share how the bank account data looks like when you try to issue the card on a new policy? Is the VBBA added for the first policy considered as eligible (is it OPEN state). Maybe that is why it's showing the "continue setup" as the bank account is not in the open state ready fro reimbursements

**Comment by koko57** [comment]
@mountiny no I think the problem here is with what I mentioned earlier - shouldShowContinueSetupButton is set using the data for the previous policy, but as in URL we have 'new' step the OpenReimbursementAccountPage is fired and the response overwrites the old data - we no longer have full achData for the reimbursement account. I think it would be good to add a check if the policyID from achData matches policyID from the URL - if not we can start a fresh new setup flow.

But now a question - is it ok that we're clearing the reimbursementAccount data?

**Comment by mountiny** [comment]
From what I understand, the `REIMBURSEMENT_ACCOUNT` key is used during the bank account setup, right that suggests it makes sense that it's overwritten if we enter the new flow. 

We had this mock 
<img width="774" alt="image" src="https://github.com/user-attachments/assets/558b2a2f-8e1c-48c9-93f3-41efa9cf3c29">
but in that case, it's about the same workspace.

In this bug, we are discussing two different workspaces, so we are adding the bank account to the new workspace.

So in that case it makes sense to clear the flow and start over

**Comment by koko57** [comment]
Overall whatever we will decide on this edge case I think that adding the check for the policyID will solve the problem, wdyt @trjExpensify @mountiny @grgia?

**Comment by mountiny** [comment]
> I think it would be good to add a check if the policyID from achData matches policyID from the URL - if not we can start a fresh new setup flow.

Lets try that and test out if we find some regressions

**Comment by koko57** [comment]
I'm working on implementing the logic here - I thought that this check would be an one-liner, but currently used logic is more convoluted than I thought and I need to take care of other things to make this work properly.

**Comment by koko57** [comment]
Ok, so it's actually nice - if we start the flow go to the other workspace the data for the reimbursement account will be overwritten, but if we go back to the first workspace we will get the correct data for what we left off for this workspace thanks to `OpenReimbursementAccountPage` call. I still need to tinker a bit with conditions and initial setup for `ReimbursementAccountPage` but I'm getting close to opening the PR for this issue.

**Comment by koko57** [comment]
I just need a confirmation here - is it ok that for a newly created Workspace `OpenReimbursementAccountPage` returns 'plaid' as a substep?
<img width="1346" alt="Screenshot 2024-09-25 at 10 19 05" src="https://github.com/user-attachments/assets/34d84a93-5df3-4e11-a74b-c2a948a7dd55">

**Comment by koko57** [comment]
Seems like I need to rewrite `ReimbursementAccountPage` logic a bit to suit our current needs, for now the changes I thought would be sufficient don't test well in all the cases.

**Comment by koko57** [comment]
One question though - I started the VBBA flow with Plaid then left it at Personal Info step (What's your legal name?), when I returned to the flow I got this response from `OpenReimbursementAccountPage` - the substep was "manual" and the currentStep was "BankAccountStep" although before leaving the flow it was saved as "plaid" and "RequestorStep" in reimbursementAccount. Is this expected?
<img width="1492" alt="Screenshot 2024-09-25 at 14 29 32" src="https://github.com/user-attachments/assets/141faa1d-c301-4106-8d1c-54ac332dbd6b">

**Comment by mountiny** [comment]
Does not sound expected to me @nkuoch do you know if this should have been saved after selecting the plaid option?

**Comment by koko57** [comment]
@mountiny @nkuoch I will also check if we're sending the request properly - I've noticed today that we're sometimes sending it with the old data.

**Comment by koko57** [comment]
<img width="1491" alt="Screenshot 2024-09-25 at 17 31 15" src="https://github.com/user-attachments/assets/176368a1-b67c-4c56-a6a2-878e67cc4901">
ok, we're sending BankAccountStep as a step (should we do that? shouldn't we send RequestorStep instead if we've finished the BankAccountStep earlier?)
but we're also sending "plaid" subStep but we get "manual" in the response
<img width="1492" alt="Screenshot 2024-09-25 at 17 31 32" src="https://github.com/user-attachments/assets/314d6ce3-e32b-4ec6-a25e-d92852c6704e">

**Comment by trjExpensify** [comment]
Can we bring this convo to the thread to align faster and keep this moving?

**Comment by trjExpensify** [comment]
It seems very strange to be changing to `manual` when `plaid` was initially chosen.

**Comment by nkuoch** [comment]
I believe it's not a bug. It's because when coming back, I believe we have no way to display the plaid selection without them trying to log in back to Plaid. So we just show the user the account numbers retrieved from Plaid. Also, even if it's shown as manual, logic wise back end side, we still consider it as added as Plaid, as we still have the plaidAccessToken

**Comment by koko57** [comment]
ok, but what if the user changes the numbers in the manual page? And if this step was done what's the point to show BankAccountStep when returning to flow? Maybe we could show the user the step from what they left off (RequestorStep) but enable them to go back to BankAccountStep with manual setup flow?

**Comment by nkuoch** [comment]
They shouldn't be able to edit the numbers. Fields should be disabled. And we also prevent it backend side

**Comment by nkuoch** [comment]
> Maybe we could show the user the step from what they left off (RequestorStep) but enable them to go back to BankAccountStep with manual setup flow?

I'll check - No need to block on it though.

**Comment by koko57** [comment]
> They shouldn't be able to edit the numbers. Fields should be disabled. And we also prevent it backend side

Ok, thanks, so I'm adding another thing to the todo list - for now we have the fields enabled.

**Comment by koko57** [comment]
I think I finally have the working solution, but now I need to test it thoroughly and disable the inputs for the manual step that Nathalie mentioned

**Comment by koko57** [comment]
I've found another problem today - when I start the flow for Workspace A, leave it at, let's say company step (Personal Info aka RequestorStep all filled - in) and then I start the flow for the Workspace B (no VBBA before) - After I get through plaid step, I choose the same bank account and I proceed to Personal Info step I get all the data from the previous workspace. Is this intentional?
<img width="595" alt="Screenshot 2024-09-26 at 18 12 42" src="https://github.com/user-attachments/assets/ede69e51-9dcd-4c45-9e02-9dff0ffd3659">

**Comment by nkuoch** [comment]
yes it's intentional

**Comment by koko57** [comment]
I think I've finally got it right 😭 PR ready for review https://github.com/Expensify/App/pull/49687

**Comment by CortneyOfstad** [comment]
Relinking Job post here — https://www.upwork.com/ab/applicants/1836823617385889668/job-details — since it got buried above

**Comment by DylanDylann** [comment]
@CortneyOfstad The checklist isn't required here

@mountiny Is the payment handled on the project or independently here?

**Comment by CortneyOfstad** [comment]
@strepanier03 I am heading OoO (10/15 to 10/23), so got this reassigned to keep it moving! Payment is due 10/16 and @DylanDylann is owed $250 👍 

Per the comment [here](https://github.com/Expensify/App/issues/49447#issuecomment-2412696919) — checklist is not needed!

**Comment by mountiny** [comment]
We can handle this separately

**Comment by mountiny** [comment]
250 to @DylanDylann

**Comment by DylanDylann** [comment]
@CortneyOfstad Bump on payment

**Comment by strepanier03** [comment]
Finished up the payment for @DylanDylann. Closing out 👋

**Comment by tgolen** [comment]
Reopening for @DylanDylann and @CortneyOfstad to complete the regression checklist.

**Comment by DylanDylann** [comment]
@mountiny Did we include this regression test on project overall?

**Comment by mountiny** [comment]
I think this is a niche flow that would not make sense to create a regression test for Applause to run through with every test. Can you complete the rest of the checklist please?

**Comment by CortneyOfstad** [comment]
I'm back from OoO (thanks @strepanier03 for holding down the fort!)

**Comment by DylanDylann** [comment]
BugZero Checklist: The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

 [@DylanDylann] The PR that introduced the bug has been identified. Link to the PR: **It more like a new improvement**
 [@DylanDylann] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: **NA**
 [@DylanDylann] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: **NA**
 [@DylanDylann] Determine if we should create a regression test for this bug. **No**
 [@DylanDylann] If we decide to create a [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) for the bug, please propose the regression test steps to ensure the same bug will not reach production again.






Do we agree 👍 or 👎

**Comment by CortneyOfstad** [comment]
Looks good @DylanDylann! Thanks!
